lib/block_view_lockups.go

```diff
@@ -0,0 +1,1784 @@
+	}
+
+	// If this is a DeSo lockup, ensure the amount is less than 2**64.
+	if txMeta.ProfilePublicKey.IsZeroPublicKey() {
+		maxUint64, _ := uint256.FromBig(big.NewInt(0).SetUint64(math.MaxUint64))
+		if txMeta.LockedCoinsToTransferBaseUnits.Gt(maxUint64) {
+			return 0, 0, nil, errors.Wrap(RuleErrorCoinLockupTransferOfDeSoCausesOverflow,
+				"_connectCoinLockupTransfer")
+		}
+	}
+
+	// Fetch PKIDs for the recipient and sender.
+	senderPKIDEntry := bav.GetPKIDForPublicKey(txn.PublicKey)
+	senderPKID := senderPKIDEntry.PKID
+	receiverPKIDEntry := bav.GetPKIDForPublicKey(txMeta.RecipientPublicKey.ToBytes())
+	receiverPKID := receiverPKIDEntry.PKID
+	profilePKIDEntry := bav.GetPKIDForPublicKey(txMeta.ProfilePublicKey.ToBytes())
+	profilePKID := profilePKIDEntry.PKID
+
+	// Ensure the sender and receiver are different.
+	if senderPKID.Eq(receiverPKID) {
+		return 0, 0, nil, errors.Wrapf(RuleErrorCoinLockupTransferSenderEqualsReceiver,
+			"_connectCoinLockupTransfer")
+	}
+
+	// Verify the transfer restrictions attached to the transfer.
+	profileEntry := bav.GetProfileEntryForPKID(profilePKID)
+	if profileEntry.DAOCoinEntry.LockupTransferRestrictionStatus == TransferRestrictionStatusProfileOwnerOnly &&
+		!profilePKID.Eq(senderPKID) {
+		return 0, 0, nil, errors.Wrapf(RuleErrorCoinLockupTransferRestrictedToProfileOwner,
+			"_connectCoinLockupTransfer")
+	}
+	if profileEntry.DAOCoinEntry.LockupTransferRestrictionStatus == TransferRestrictionStatusDAOMembersOnly {
+		// TODO: Determine if this is desired behavior. We assume the sender must be part of the DAO to have
+		//       transferable coins. It seems weird to tie locked DAO coin transfers to unlocked DAO coin balances.
+		//       An alternative approach is not allow the "TransferRestrictionStatusDAOMembersOnly" restriction.
+		receiverBalanceEntry := bav.GetBalanceEntry(receiverPKID, profilePKID, true)
+		if receiverBalanceEntry.BalanceNanos.IsZero() {
+			return 0, 0, nil,
+				errors.Wrapf(RuleErrorCoinLockupTransferRestrictedToDAOMembers, "_connectCoinLockupTransfer")
+		}
+	}
+
+	// Fetch the sender's balance entries.
+	senderLockedBalanceEntry := bav.GetLockedBalanceEntryForHODLerPKIDProfilePKIDExpirationTimestampNanoSecs(
+		senderPKID, profilePKID, txMeta.ExpirationTimestampUnixNanoSecs)
+	prevSenderLockedBalanceEntry := senderLockedBalanceEntry.Copy()
+
+	// Check that the sender's balance entry has sufficient balance.
+	if txMeta.LockedCoinsToTransferBaseUnits.Gt(&senderLockedBalanceEntry.BalanceBaseUnits) {
+		return 0, 0, nil, errors.Wrapf(RuleErrorCoinLockupTransferInsufficientBalance,
+			"_connectCoinLockupTransfer")
+	}
+
+	// Credit the sender's balance entry.
+	senderLockedBalanceEntry.BalanceBaseUnits = *uint256.NewInt().Sub(
+		&senderLockedBalanceEntry.BalanceBaseUnits, txMeta.LockedCoinsToTransferBaseUnits)
+
+	// Fetch the recipient's balance entry.
+	receiverLockedBalanceEntry := bav.GetLockedBalanceEntryForHODLerPKIDProfilePKIDExpirationTimestampNanoSecs(
```

Comment:
This is tricky. It works because the db function returns an empty balance entry constructed for you as a convenience. Leave a comment here. Or, honestly, I think it would make it more clear if the GetLocked...() function returned nil and you constructed it here explicitly.